tests/integration/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+#![allow(unused_imports)]
+
+use std::{collections::HashMap, num::NonZeroUsize, thread::sleep, time::Duration};
+
+use common::{
+    async_interface::ReqResPair,
+    gm_request_response::{Request, Transaction, TxStatus},
+    meta_exchange::Response,
+    range_key::{KeyOffset, KeyRange, TransactionID},
+};
+use gateway::core::Gateway;
+use maroon::{
+    app::{App, CurrentOffsets, Params, Request as AppRequest, Response as AppResponse},
+    stack,
+};
+use tokio::sync::oneshot;
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_some() {
+    env_logger::init();
+
+    let (_shutdown_tx_0, shutdown_rx_0) = oneshot::channel();
+    let (_shutdown_tx_1, shutdown_rx_1) = oneshot::channel();
+    let (_shutdown_tx_2, shutdown_rx_2) = oneshot::channel();
+
+    let params = Params {
+        advertise_period: Duration::from_millis(500),
+        consensus_nodes: NonZeroUsize::new(2).unwrap(),
+    };
+
+    // create nodes and gateway
+
+    let mut node0 = stack::create_stack(
+        vec![
+            "/ip4/127.0.0.1/tcp/3001".to_string(),
+            "/ip4/127.0.0.1/tcp/3002".to_string(),
+        ],
+        "/ip4/0.0.0.0/tcp/3000".to_string(),
+        params.clone(),
+    )
+    .unwrap();
+    let mut node1 = stack::create_stack(
+        vec![
+            "/dns4/localhost/tcp/3000".to_string(),
+            "/dns4/localhost/tcp/3002".to_string(),
+        ],
+        "/ip4/0.0.0.0/tcp/3001".to_string(),
+        params.clone(),
+    )
+    .unwrap();
+    let mut node2 = stack::create_stack(
+        vec![
+            "/ip4/127.0.0.1/tcp/3000".to_string(),
+            "/ip4/127.0.0.1/tcp/3001".to_string(),
+        ],
+        "/ip4/0.0.0.0/tcp/3002".to_string(),
+        params,
+    )
+    .unwrap();
+
+    let mut gw = Gateway::new(vec![
+        "/ip4/127.0.0.1/tcp/3000".to_string(),
+        "/ip4/127.0.0.1/tcp/3001".to_string(),
+        "/ip4/127.0.0.1/tcp/3002".to_string(),
+    ])
+    .unwrap();
+
+    let mut state_check_0 = node0.get_state_interface();
+    let mut state_check_1 = node1.get_state_interface();
+    let mut state_check_2 = node2.get_state_interface();
+
+    // run nodes and gateway
+
+    tokio::spawn(async move { node0.loop_until_shutdown(shutdown_rx_0).await });
+    tokio::spawn(async move { node1.loop_until_shutdown(shutdown_rx_1).await });
+    tokio::spawn(async move { node2.loop_until_shutdown(shutdown_rx_2).await });
+
+    gw.start_on_background().await;
+
+    // wait until they are connected
+    tokio::time::sleep(Duration::from_secs(1)).await;
```

Comment:
Meh =) although the above reads nicely, except the part it's not clear how do you check if the nodes were indeed shut down.

(I.e. I would expect the results of `tokio::spawn(...)` to be collected somewhere, so that at the end of the test you would call something like `assert_teared_down_within(from_millis(1000))`, since you'd order the nodes to shut down by that time.)